File: boot/acp-spring-boot-starter/src/main/java/pers/acp/springboot/socket/udp/UdpServer.java
package pers.acp.springboot.socket.udp;

import io.netty.bootstrap.Bootstrap;
import io.netty.channel.*;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioDatagramChannel;
import pers.acp.core.interfaces.IDaemonService;
import pers.acp.core.log.LogFactory;
import pers.acp.springboot.conf.SocketListenerConfiguration;
import pers.acp.springboot.socket.base.ISocketServerHandle;

public final class UdpServer implements IDaemonService, Runnable {

    private final LogFactory log = LogFactory.getInstance(this.getClass());

    private int port;

    private EventLoopGroup bossGroup;

    private SocketListenerConfiguration socketListenerConfiguration;

    private ISocketServerHandle socketServerHandle;

    /**
     * 构造函数
     *
     * @param port                        端口
     * @param socketListenerConfiguration 监听服务配置
     * @param socketServerHandle          接收报文处理对象
     */
    public UdpServer(int port, SocketListenerConfiguration socketListenerConfiguration, ISocketServerHandle socketServerHandle) {
        this.port = port;
        this.bossGroup = new NioEventLoopGroup();
        this.socketListenerConfiguration = socketListenerConfiguration;
        this.socketServerHandle = socketServerHandle;
    }

    @Override
    public void run() {
        if (socketServerHandle != null) {
            Bootstrap bootstrap = new Bootstrap();
            bootstrap.group(bossGroup)
                    .channel(NioDatagramChannel.class)
                    .option(ChannelOption.SO_BROADCAST, true)
                    .handler(new ChannelInitializer<NioDatagramChannel>() {
                        @Override
                        protected void initChannel(NioDatagramChannel ch) {
                            ch.pipeline().addLast(new UdpServerHandle(socketListenerConfiguration, socketServerHandle));
                        }
                    });
            try {
                bootstrap.bind(port).sync().channel().closeFuture().sync();
            } catch (Exception e) {
                log.error(e.getMessage(), e);
            }
        } else {
            log.error("udp listen server is stop,case by:response object is null[BaseSocketHandle]");
        }
    }

    @Override
    public String getServiceName() {
        return socketListenerConfiguration.getName();
    }

    @Override
    public void stopService() {
        bossGroup.shutdownGracefully();
    }

}

File: boot/acp-spring-boot-starter/src/main/kotlin/pers/acp/spring/boot/socket/udp/UdpServer.kt
package pers.acp.spring.boot.socket.udp

import io.netty.bootstrap.Bootstrap
import io.netty.channel.*
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioDatagramChannel
import pers.acp.core.interfaces.IDaemonService
import pers.acp.core.log.LogFactory
import pers.acp.spring.boot.conf.SocketListenerConfiguration
import pers.acp.spring.boot.interfaces.LogAdapter
import pers.acp.spring.boot.socket.base.ISocketServerHandle

/**
 * Udp 服务端
 */
class UdpServer
/**
 * 构造函数
 *
 * @param log                         日志适配器
 * @param port                        端口
 * @param socketListenerConfiguration 监听服务配置
 * @param socketServerHandle          接收报文处理对象
 */
(private val log: LogAdapter,
 private val port: Int,
 private val socketListenerConfiguration: SocketListenerConfiguration,
 private val socketServerHandle: ISocketServerHandle?) : IDaemonService, Runnable {

    private val bossGroup: EventLoopGroup

    init {
        this.bossGroup = NioEventLoopGroup()
    }

    override fun run() {
        if (socketServerHandle != null) {
            try {
                Bootstrap().group(bossGroup)
                        .channel(NioDatagramChannel::class.java)
                        .option(ChannelOption.SO_BROADCAST, true)
                        .handler(object : ChannelInitializer<NioDatagramChannel>() {
                            override fun initChannel(ch: NioDatagramChannel) {
                                ch.pipeline().addLast(UdpServerHandle(log, socketListenerConfiguration, socketServerHandle))
                            }
                        }).bind(port).sync().channel().closeFuture().sync()
            } catch (e: Exception) {
                log.error(e.message, e)
            }

        } else {
            log.error("udp listen server is stop,case by:response object is null[BaseSocketHandle]")
        }
    }

    override fun getServiceName(): String = socketListenerConfiguration.name

    override fun stopService() {
        bossGroup.shutdownGracefully()
    }

}
